sys/arm64/arm64/efirt_support.S

```diff
@@ -0,0 +1,159 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <machine/asm.h>
+#include <machine/armreg.h>
+
+#include "assym.inc"
+
+#ifdef __CHERI_PURE_CAPABILITY__
+/*
+ * int efi_rt_arch_call_nofault(struct efirt_callinfo *ec)
+ *
+ * Wrap EFI calls to save caller-saved capability registers until we have a
+ * pure-capability interface.
+ *
+ * NB: This temporarily installs an almighty DDC.
+ */
+ENTRY(efi_rt_arch_call_nofault)
+	sub	csp, csp, #(17 * PTR_WIDTH)
+
+	/*
+	 * Save all capability registers that should be callee-saved, plus save
+	 * the efirt_callinfo pointer so we can set ec_efi_status on return.
+	 */
+	stp	c16, c17, [csp]
+	stp	c18, c19, [csp, #(2 * PTR_WIDTH)]
+	stp	c20, c21, [csp, #(4 * PTR_WIDTH)]
+	stp	c22, c23, [csp, #(6 * PTR_WIDTH)]
+	stp	c24, c25, [csp, #(8 * PTR_WIDTH)]
+	stp	c26, c27, [csp, #(10 * PTR_WIDTH)]
+	stp	c28, c29, [csp, #(12 * PTR_WIDTH)]
+	stp	c30,  c0, [csp, #(14 * PTR_WIDTH)]
+
+	/* Save CSP as the metadata will be clobbered by the EFI call */
+	mov	c8, csp
+	str	c8, [csp, #(16 * PTR_WIDTH)]
+
+	/* Get the number of arguments for the EFI call */
+	ldr	x8, [c0, #EC_ARGCNT]
+
+	/* Check we have no more than 5 arguments */
+	cmp	x8, #5
+	b.hi	1f
+
+	/* Get target function pointer */
+	ldr	c6, [c0, #EC_FPTR]
+
+	/* Jump to 0f + 4 * (5 - ec_argcnt) + 1 */
+	adr	c9, 0f + 1
+	mov	x10, #20
+	sub	x8, x10, x8, lsl #2
+	add	c9, c9, x8, uxtx
+	br	c9
+
+	/* Load the arguments into registers */
+0:	ldr	x4, [c0, #EC_ARG5]
+	ldr	x3, [c0, #EC_ARG4]
+	ldr	x2, [c0, #EC_ARG3]
+	ldr	x1, [c0, #EC_ARG2]
+	ldr	x0, [c0, #EC_ARG1]
+
+	/* Install almighty DDC just for EFI (and our CSP restore) */
+	adrp	c7, :got:kernel_root_cap
+	ldr	c7, [c7, :got_lo12:kernel_root_cap]
+	ldr	c7, [c7]
+	msr	ddc, c7
+
+	/*
+	 * Temporarily disable all interrupts as we can't handle them whilst
+	 * CSP is clobbered during the EFI call.
+	 */
+	mrs	x16, daif
+	msr	daifset, #(DAIF_ALL)
+
+	/*
+	 * Need to be in A64 so LR has the LSB clear, since the callee's
+	 * RET X30 won't honour or clear the LSB.
+	 */
+	bx	#4
+	.arch_extension	noc64
+	.arch_extension	a64c
+	blr	c6
+	bx	#4
+	.arch_extension	c64
+
+	/* EFI calls run in the same EL so trash SP metadata */
+	ldr	c8, [sp, #(16 * PTR_WIDTH)]
```

Comment:
I hate doing this specifically but I don't see a better option...

Reply:
I assume you mean having to depend on a DDC load to restore csp?  I agree that I don't see a better option if so.